plugins/providers/alicloudiam/provider.go

```diff
@@ -0,0 +1,288 @@
+package alicloudiam
+
+import (
+	"fmt"
+	ram "github.com/alibabacloud-go/ram-20150501/v2/client"
+	"github.com/goto/guardian/core/provider"
+	"github.com/goto/guardian/domain"
+	"github.com/goto/guardian/pkg/log"
+	"github.com/goto/guardian/utils"
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/net/context"
+	"regexp"
+	"strings"
+)
+
+//go:generate mockery --name=AliCloudIamClient --exported --with-expecter
+type AliCloudIamClient interface {
+	GrantAccess(ctx context.Context, policyName, policyType, username string) error
+	RevokeAccess(ctx context.Context, policyName, policyType, username string) error
+	GrantAccessToRole(ctx context.Context, policyName, policyType, roleName string) error
+	RevokeAccessFromRole(ctx context.Context, policyName, policyType, roleName string) error
+	ListAccess(ctx context.Context, pc domain.ProviderConfig, resources []*domain.Resource) (domain.MapResourceAccess, error)
+	GetAllPoliciesByType(_ context.Context, policyType string, maxItems int32) ([]*ram.ListPoliciesResponseBodyPoliciesPolicy, error)
+}
+
+//go:generate mockery --name=encryptor --exported --with-expecter
+type encryptor interface {
+	domain.Crypto
+}
+
+type Provider struct {
+	provider.PermissionManager
+	provider.UnimplementedClient
+
+	typeName string
+	Clients  map[string]AliCloudIamClient
+	crypto   encryptor
+	logger   log.Logger
+}
+
+func NewProvider(typeName string, crypto encryptor, logger log.Logger) *Provider {
+	return &Provider{
+		typeName: typeName,
+		Clients:  map[string]AliCloudIamClient{},
+		crypto:   crypto,
+		logger:   logger,
+	}
+}
+
+func (p *Provider) GetType() string {
+	return p.typeName
+}
+
+func (p *Provider) CreateConfig(pc *domain.ProviderConfig) error {
+	c := NewConfig(pc, p.crypto)
+
+	if err := c.ParseAndValidate(); err != nil {
+		return err
+	}
+
+	client, err := p.getIamClient(pc)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range c.ProviderConfig.Resources {
+		if err = c.validatePermissions(r, client); err != nil {
+			return err
+		}
+	}
+
+	return c.EncryptCredentials()
+}
+
+func (p *Provider) GetResources(_ context.Context, pc *domain.ProviderConfig) ([]*domain.Resource, error) {
```

Comment:
We can remove context as the param since it is not used.

Reply:
ack!

Reply:
We cannot remove it. It required to implement this interface:
https://github.com/goto/guardian/blob/main/plugins/providers/client.go

And that interface will be used on initializing new providers at:
https://github.com/goto/guardian/blob/main/internal/server/services.go#L114

Reply:
This was likely added to encourage implementing requests with a context. By doing so, if a request is sent to a provider, it won’t block indefinitely and will automatically cancel all ongoing processes when the parent context expires or is canceled.

Reply:
But since Ali SDK has not support for it, I already raised the question here:
https://github.com/aliyun/darabonba-openapi/issues/171